### ISSUE TITLE
Onboarding startup issues

### ISSUE THREAD

**Issue by gr4per** [OpusCapita/onboarding#208]
Observer this in compose today:
```
{"serviceName":"db-init","serviceInstanceId":0,"correlationId":null,"userId":null,"requestUri":null,"timestamp":"2017-07-06T14:59:03.718Z","level":"info","message":"Initializing database... Host: 172.19.0.4, Port: 3306"}
{"serviceName":"auth","serviceInstanceId":0,"correlationId":null,"userId":null,"requestUri":null,"timestamp":"2017-07-06T14:59:03.732Z","level":"error","message":"{ Error: operation timed out after 25640 ms, 50 tries with error: connect ECONNREFUSED 172.19.0.4:3306\n    at /var/tmp/base/node_modules/bluebird-retry/lib/bluebird-retry.js:109:35\n    at runCallback (timers.js:672:20)\n    at tryOnImmediate (timers.js:645:5)\n    at processImmediate [as _immediateCallback] (timers.js:617:5)\nFrom previous event:\n    at Promise.then (/var/tmp/base/node_modules/sequelize/lib/promise.js:21:17)\n    at /var/tmp/base/node_modules/bluebird-retry/lib/bluebird-retry.js:118:53\n    at runCallback (timers.js:672:20)\n    at tryOnImmediate (timers.js:645:5)\n    at processImmediate [as _immediateCallback] (timers.js:617:5)\n  failure: \n   { SequelizeConnectionRefusedError: connect ECONNREFUSED 172.19.0.4:3306\n       at Handshake._callback (/var/tmp/base/node_modules/sequelize/lib/dialects/mysql/connection-manager.js:80:20)\n       at Handshake.Sequence.end (/var/tmp/base/node_modules/mysql/lib/protocol/sequences/Sequence.js:86:24)\n       at Protocol.handleNetworkError (/var/tmp/base/node_modules/mysql/lib/protocol/Protocol.js:364:14)\n       at Connection._handleNetworkError (/var/tmp/base/node_modules/mysql/lib/Connection.js:428:18)\n       at emitOne (events.js:96:13)\n       at Socket.emit (events.js:188:7)\n       at emitErrorNT (net.js:1277:8)\n       at _combinedTickCallback (internal/process/next_tick.js:80:11)\n       at process._tickCallback (internal/process/next_tick.js:104:9)\n   From previous event:\n       at ConnectionManager.connect (/var/tmp/base/node_modules/sequelize/lib/dialects/mysql/connection-manager.js:55:10)\n       at ConnectionManager.<anonymous> (/var/tmp/base/node_modules/sequelize/lib/dialects/abstract/connection-manager.js:247:43)\n       at runCallback (timers.js:672:20)\n       at tryOnImmediate (timers.js:645:5)\n       at processImmediate [as _immediateCallback] (timers.js:617:5)\n   From previous event:\n       at Promise.then (/var/tmp/base/node_modules/sequelize/lib/promise.js:21:17)\n       at ConnectionManager.$connect (/var/tmp/base/node_modules/sequelize/lib/dialects/abstract/connection-manager.js:246:70)\n       at ConnectionManager.getConnection (/var/tmp/base/node_modules/sequelize/lib/dialects/abstract/connection-manager.js:203:44)\n       at Sequelize.query (/var/tmp/base/node_modules/sequelize/lib/sequelize.js:844:83)\n       at Sequelize.authenticate (/var/tmp/base/node_modules/sequelize/lib/sequelize.js:1085:15)\n       at Promise (/var/tmp/base/node_modules/ocbesbn-db-init/index.js:204:27)\n   From previous event:\n       at task (/var/tmp/base/node_modules/ocbesbn-db-init/index.js:192:26)\n       at /var/tmp/base/node_modules/bluebird-retry/lib/bluebird-retry.js:71:29\n   From previous event:\n       at try_once (/var/tmp/base/node_modules/bluebird-retry/lib/bluebird-retry.js:70:24)\n       at ontimeout (timers.js:386:14)\n       at tryOnTimeout (timers.js:250:5)\n       at Timer.listOnTimeout (timers.js:214:5)\n     name: 'SequelizeConnectionRefusedError',\n     message: 'connect ECONNREFUSED 172.19.0.4:3306',\n     parent: \n      { Error: connect ECONNREFUSED 172.19.0.4:3306\n          at Object.exports._errnoException (util.js:1018:11)\n          at exports._exceptionWithHostPort (util.js:1041:20)\n          at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1086:14)\n          --------------------\n          at Protocol._enqueue (/var/tmp/base/node_modules/mysql/lib/protocol/Protocol.js:141:48)\n          at Protocol.handshake (/var/tmp/base/node_modules/mysql/lib/protocol/Protocol.js:52:41)\n          at Connection.connect (/var/tmp/base/node_modules/mysql/lib/Connection.js:130:18)\n          at /var/tmp/base/node_modules/sequelize/lib/dialects/mysql/connection-manager.js:75:16\n          at Promise._execute (/var/tmp/base/node_modules/bluebird/js/release/debuggability.js:300:9)\n          at Promise._resolveFromExecutor (/var/tmp/base/node_modules/bluebird/js/release/promise.js:483:18)\n          at new Promise (/var/tmp/base/node_modules/bluebird/js/release/promise.js:79:10)\n          at ConnectionManager.connect (/var/tmp/base/node_modules/sequelize/lib/dialects/mysql/connection-manager.js:55:10)\n          at ConnectionManager.<anonymous> (/var/tmp/base/node_modules/sequelize/lib/dialects/abstract/connection-manager.js:247:43)\n          at ConnectionManager.tryCatcher (/var/tmp/base/node_modules/bluebird/js/release/util.js:16:23)\n          at Promise._settlePromiseFromHandler (/var/tmp/base/node_modules/bluebird/js/release/promise.js:512:31)\n          at Promise._settlePromise (/var/tmp/base/node_modules/bluebird/js/release/promise.js:569:18)\n          at Promise._settlePromise0 (/var/tmp/base/node_modules/bluebird/js/release/promise.js:614:10)\n          at Promise._settlePromises (/var/tmp/base/node_modules/bluebird/js/release/promise.js:693:18)\n          at Async._drainQueue (/var/tmp/base/node_modules/bluebird/js/release/async.js:133:16)\n          at Async._drainQueues (/var/tmp/base/node_modules/bluebird/js/release/async.js:143:10)\n          at Immediate.Async.drainQueues (/var/tmp/base/node_modules/bluebird/js/release/async.js:17:14)\n        code: 'ECONNREFUSED',\n        errno: 'ECONNREFUSED',\n        syscall: 'connect',\n        address: '172.19.0.4',\n        port: 3306,\n        fatal: true },\n     original: \n      { Error: connect ECONNREFUSED 172.19.0.4:3306\n          at Object.exports._errnoException (util.js:1018:11)\n          at exports._exceptionWithHostPort (util.js:1041:20)\n          at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1086:14)\n          --------------------\n          at Protocol._enqueue (/var/tmp/base/node_modules/mysql/lib/protocol/Protocol.js:141:48)\n          at Protocol.handshake (/var/tmp/base/node_modules/mysql/lib/protocol/Protocol.js:52:41)\n          at Connection.connect (/var/tmp/base/node_modules/mysql/lib/Connection.js:130:18)\n          at /var/tmp/base/node_modules/sequelize/lib/dialects/mysql/connection-manager.js:75:16\n          at Promise._execute (/var/tmp/base/node_modules/bluebird/js/release/debuggability.js:300:9)\n          at Promise._resolveFromExecutor (/var/tmp/base/node_modules/bluebird/js/release/promise.js:483:18)\n          at new Promise (/var/tmp/base/node_modules/bluebird/js/release/promise.js:79:10)\n          at ConnectionManager.connect (/var/tmp/base/node_modules/sequelize/lib/dialects/mysql/connection-manager.js:55:10)\n          at ConnectionManager.<anonymous> (/var/tmp/base/node_modules/sequelize/lib/dialects/abstract/connection-manager.js:247:43)\n          at ConnectionManager.tryCatcher (/var/tmp/base/node_modules/bluebird/js/release/util.js:16:23)\n          at Promise._settlePromiseFromHandler (/var/tmp/base/node_modules/bluebird/js/release/promise.js:512:31)\n          at Promise._settlePromise (/var/tmp/base/node_modules/bluebird/js/release/promise.js:569:18)\n          at Promise._settlePromise0 (/var/tmp/base/node_modules/bluebird/js/release/promise.js:614:10)\n          at Promise._settlePromises (/var/tmp/base/node_modules/bluebird/js/release/promise.js:693:18)\n          at Async._drainQueue (/var/tmp/base/node_modules/bluebird/js/release/async.js:133:16)\n          at Async._drainQueues (/var/tmp/base/node_modules/bluebird/js/release/async.js:143:10)\n          at Immediate.Async.drainQueues (/var/tmp/base/node_modules/bluebird/js/release/async.js:17:14)\n        code: 'ECONNREFUSED',\n        errno: 'ECONNREFUSED',\n        syscall: 'connect',\n        address: '172.19.0.4',\n        port: 3306,\n        fatal: true } },\n  code: 'ETIMEDOUT' }"}
[nodemon] app crashed - waiting for file changes before starting...
```
This was auth service failing to connect to db and hence crashing, causing kong error when trying to get to onboarding.

Seems to be a problem with udn as i cannot connect 172.19.0.4:3306 from within auth container.
This is the ip and port consul returns for mysql.
However, mysql is up and running on the 172.19.0.1 (gateway=docker host), so the published ports are accessible.

Restarting the auth container shows different behaviour and the udp is properly routed.
Observed on docker 1.13.0

**Comment by gr4per** [comment]
Another try with this error
```
{"serviceName":"api-registrator","serviceInstanceId":0,"correlationId":null,"userId":null,"requestUri":null,"timestamp":"2017-07-06T20:39:17.952Z","level":"info","message":"{ Error: operation timed out after 45896 ms, 10 tries with error: The requested endpoint could not be found or did not pass health checks: kong\n    at /home/node/service-register/node_modules/bluebird-retry/lib/bluebird-retry.js:109:35\n    at tryCatcher (/home/node/service-register/node_modules/bluebird/js/release/util.js:16:23)\n    at /home/node/service-register/node_modules/bluebird/js/release/catch_filter.js:24:41\n    at tryCatcher (/home/node/service-register/node_modules/bluebird/js/release/util.js:16:23)\n    at Promise._settlePromiseFromHandler (/home/node/service-register/node_modules/bluebird/js/release/promise.js:512:31)\n    at Promise._settlePromise (/home/node/service-register/node_modules/bluebird/js/release/promise.js:569:18)\n    at Promise._settlePromise0 (/home/node/service-register/node_modules/bluebird/js/release/promise.js:614:10)\n    at Promise._settlePromises (/home/node/service-register/node_modules/bluebird/js/release/promise.js:689:18)\n    at Async._drainQueue (/home/node/service-register/node_modules/bluebird/js/release/async.js:133:16)\n    at Async._drainQueues (/home/node/service-register/node_modules/bluebird/js/release/async.js:143:10)\n    at Immediate.Async.drainQueues (/home/node/service-register/node_modules/bluebird/js/release/async.js:17:14)\n    at runCallback (timers.js:637:20)\n    at tryOnImmediate (timers.js:610:5)\n    at processImmediate [as _immediateCallback] (timers.js:582:5)\n  failure: \n   Error: The requested endpoint could not be found or did not pass health checks: kong\n       at Promise.all.then.results (/home/node/service-register/node_modules/ocbesbn-config/index.js:290:13)\n       at tryCatcher (/home/node/service-register/node_modules/bluebird/js/release/util.js:16:23)\n       at Promise._settlePromiseFromHandler (/home/node/service-register/node_modules/bluebird/js/release/promise.js:512:31)\n       at Promise._settlePromise (/home/node/service-register/node_modules/bluebird/js/release/promise.js:569:18)\n       at Promise._settlePromise0 (/home/node/service-register/node_modules/bluebird/js/release/promise.js:614:10)\n       at Promise._settlePromises (/home/node/service-register/node_modules/bluebird/js/release/promise.js:693:18)\n       at Promise._fulfill (/home/node/service-register/node_modules/bluebird/js/release/promise.js:638:18)\n       at PromiseArray._resolve (/home/node/service-register/node_modules/bluebird/js/release/promise_array.js:126:19)\n       at PromiseArray._promiseFulfilled (/home/node/service-register/node_modules/bluebird/js/release/promise_array.js:144:14)\n       at Promise._settlePromise (/home/node/service-register/node_modules/bluebird/js/release/promise.js:574:26)\n       at Promise._settlePromise0 (/home/node/service-register/node_modules/bluebird/js/release/promise.js:614:10)\n       at Promise._settlePromises (/home/node/service-register/node_modules/bluebird/js/release/promise.js:693:18)\n       at Async._drainQueue (/home/node/service-register/node_modules/bluebird/js/release/async.js:133:16)\n       at Async._drainQueues (/home/node/service-register/node_modules/bluebird/js/release/async.js:143:10)\n       at Immediate.Async.drainQueues (/home/node/service-register/node_modules/bluebird/js/release/async.js:17:14)\n       at runCallback (timers.js:637:20),\n  code: 'ETIMEDOUT' }"}
```
This was due to registrator not getting consul and bailing out
```
gr4per@dev-grapear1:~/onboarding$ docker logs a7e85972e063
2017/07/06 20:38:02 Starting registrator v7 ...
2017/07/06 20:38:02 Using consul adapter: consul://consul:8500
2017/07/06 20:38:02 Connecting to backend (0/0)
2017/07/06 20:38:02 Get http://consul:8500/v1/status/leader: dial tcp 172.18.0.2:8500: connection refused
```
Hence api-registrator not finding kong, so no apis registered

**Comment by gr4per** [comment]
Even better, now onboarding craps up with this error after a long time:
```
{"serviceName":"config","serviceInstanceId":0,"correlationId":null,"userId":null,"requestUri":null,"timestamp":"2017-07-06T20:38:34.739Z","level":"info","message":"Consul connection established."}
Unhandled rejection Error: operation timed out after 1252344 ms, 50 tries with error: operation timed out after 25070 ms, 50 tries with error: The requested endpoint could not be found or did not pass health checks: mysql
    at /home/node/onboarding/node_modules/bluebird-retry/lib/bluebird-retry.js:109:35
From previous event:
    at Promise.then (/home/node/onboarding/node_modules/sequelize/lib/promise.js:21:17)
    at try_once (/home/node/onboarding/node_modules/bluebird-retry/lib/bluebird-retry.js:81:14)
    at /home/node/onboarding/node_modules/bluebird-retry/lib/bluebird-retry.js:116:32
    at runCallback (timers.js:800:20)
    at tryOnImmediate (timers.js:762:5)
    at processImmediate [as _immediateCallback] (timers.js:733:5)
[nodemon] clean exit - waiting for changes before restart
```
This is about 20 mins and explains why we usually see Consul Connection established as last message, then nothing.
1) We should make sure retry on connection produces log output
2) this looks like a docker network stack issue, trying to upgrade version

**Comment by celber** [comment]
I am using 17.06.0-ce-mac18 (18433) (considered the same) and i can reproduce it.

**Comment by gr4per** [comment]
I have adjusted local build from compose udn to bridge network and trying that so far ok.
Reproduced issue earlier on 17.03.1
Now monitoring

Von: celber [mailto:notifications@github.com]
Gesendet: Freitag, 7. Juli 2017 15:26
An: OpusCapita/onboarding <onboarding@noreply.github.com>
Cc: Gräper Arne <Arne.Graeper@opuscapita.com>; Author <author@noreply.github.com>
Betreff: Re: [OpusCapita/onboarding] Onboarding startup issues (#208)


I am using 17.06.0-ce-mac18 (18433) (considered the same) and i can reproduce it.

**Comment by gr4per** [comment]
added networktest container to compose setup. this is additional telemetry about the network stack as it will ping main from a separate container.
I started docker-compose clean 6 times today without a single time experiencing a startup issue
So not clear what we are delaing with

**Comment by celber** [comment]
Closing as its duplicate of #201
